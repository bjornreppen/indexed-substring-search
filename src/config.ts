export const config = {
  tokenize: (phrase: string) => {
    if (!phrase) return [];
    if (phrase.indexOf(" ") < 0) {
      return [phrase];
    }
    return phrase.split(/[.,\/\:\(\) -]/);
  }
};
