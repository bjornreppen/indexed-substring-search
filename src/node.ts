export interface INode {
  [key: string]: any;
}

export interface IResult {
  key: any;
  score: number;
}
