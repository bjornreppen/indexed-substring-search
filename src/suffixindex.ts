import { config } from "./config";
import { INode } from "./node";

export class SuffixIndex {
  private index: INode;

  constructor(index: INode) {
    this.index = index;
  }

  /** Look up one or more substrings, return the set matching both */
  public queryPhrase(q: string) {
    const words = config.tokenize(q);
    const wordMatches: any = [];
    words.forEach(word => {
      wordMatches.push(this.queryWord(word));
    });
    // For our intersect, sort the shortest result set first
    wordMatches.sort((a: any, b: any) => a.length - b.length);
    const result: any = [];
    const [first, ...rest] = wordMatches;
    Object.keys(first).forEach((key: string) => {
      let score = first[key];
      for (const rx of rest) {
        if (!(key in rx)) {
          return;
        }
        score *= rx[key];
      }
      result.push({ key: key, score: score });
    });
    result.sort((a: any, b: any) => b.score - a.score);
    return result;
  }

  /** Look up a single word in the index */
  public queryWord(q: string) {
    q = q.toLowerCase();
    let current = this.index;
    for (const char of q) {
      if (!current[char]) {
        return {};
      }
      current = current[char];
    }
    const results = this.reduce(current, {});
    return results;
    //      return results.sort((a: any, b: any) => {
    //      return (Object as any).values(b)[0] - (Object as any).values(a)[0];
    //  });
  }

  private reduce(r: INode, s: any, parentKey: string = "") {
    Object.keys(r).forEach(key => {
      if (parentKey === "$") {
        s[key] = Math.max(s[key] || 0, r[key]);
      } else {
        this.reduce(r[key], s, key);
      }
    });
    return s;
  }
}
