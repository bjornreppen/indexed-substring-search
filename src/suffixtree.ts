import * as fs from "fs";
import { config } from "./config";
import { INode, IResult } from "./node";
import { SuffixIndex } from "./suffixindex";

/** A tiny substring search module
 * https://en.wikipedia.org/wiki/Suffix_tree
 * Can be optimized massively especially for space, but trading size for developer hours.
 */
export class SuffixTree {
  private index: INode;

  constructor(json = {}) {
    this.index = json;
  }

  /** Serialize the tree to string */
  public toJson() {
    return JSON.stringify(this.index);
  }

  /** Builds the qeryable index. */
  public buildIndex() {
    return new SuffixIndex(this.index);
  }

  /** index a sequence of words */
  public addSentence(sentence: string, searchResult: IResult) {
    if (!sentence) {
      return;
    }
    const words = config.tokenize(sentence);
    for (const word of words) {
      this.addWord(word, searchResult);
    }
  }

  /** index a single word */
  private addWord(word: string, searchResult: IResult) {
    if (word.length < 2) {
      return;
    }
    for (let i = 0; i < word.length; i++) {
      this.addSuffix(word.substring(i).toLowerCase(), searchResult);
    }
  }

  /** index ending of a string */
  private addSuffix(sub: string, searchResult: IResult) {
    let current = this.index;
    for (const char of sub) {
      if (!current[char]) {
        current[char] = {};
      }
      current = current[char];
    }
    if (!current.$) {
      current.$ = {};
    }
    current.$[searchResult.key] = searchResult.score;
  }
}
