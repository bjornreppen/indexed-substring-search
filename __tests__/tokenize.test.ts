import { config } from "../src/config";

test("Tokenize erosjonsmotstand", () => {
  const actual = config.tokenize("Erosjonsmotstand (i sorterte sedimenter)");
  expect(actual).toMatchSnapshot();
});
