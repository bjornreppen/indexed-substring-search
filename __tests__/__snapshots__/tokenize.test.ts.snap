// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenize erosjonsmotstand 1`] = `
Array [
  "Erosjonsmotstand",
  "i",
  "sorterte",
  "sedimenter",
]
`;
